[SNAP-2068] cleanup open transactions and TXContexts

sumwale opened the pull request:
The issue noted in SNAP-2068 is due to open transactions and TXContexts left during
cache close in case the same threads get used subsequently. This can happen for
cache reconnects for Spark executor threads for example (in the test it happens for
RMI threads used by test for method executions).

## Changes proposed in this pull request

- track hosted TXContexts for cleanup in cache close so that subsequent reconnects,
  if any, do not end up using stale TXContexts
- also rollback any open hosted TXs on TXManagerImpl close (ignoring any failures)
- moved the separate snapshotTXState ThreadLocal inside TXContext to aid in cleanup
- changed code and test calls to use TXManagerImpl.getCurrentSnapshotTXState() and
  setSnapshotTXState() instead of direct ThreadLocal handle for above change
- added removal of TXContext from global list on thread close from
  ConnectionTable.releaseThreadsSockets() that is invoked by all GemFire thread pools
- added a ThreadFactory for SnappyThriftServer for above call which should be done for
  GemFire layer cleanup in any case; also added for legacy DRDAConnThread
- also added thread stats for thrift pool like done by other GemFire pools

## Patch testing

precheckin with store

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/872

Commit:
[SNAP-2068] cleanup open transactions and TXContexts

- track hosted TXContexts for cleanup in cache close so that subsequent reconnects,
  if any, do not end up using stale TXContexts
- moved the separate snapshotTXState ThreadLocal inside TXContext to aid in cleanup
- changed code and test calls to use TXManagerImpl.getCurrentSnapshotTXState() and
  setSnapshotTXState() instead of direct ThreadLocal handle for above change
- added removal of TXContext from global list on thread close from
  ConnectionTable.releaseThreadsSockets() that is invoked by all GemFire thread pools
- added a ThreadFactory for SnappyThriftServer for above call which should be done for
  GemFire layer cleanup in any case; also added for legacy DRDAConnThread
- also added thread stats for thrift pool like done by other GemFire pools

suranjan commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXManagerImpl.java:
```diff
@@ -639,13 +646,30 @@ public final void setTXState(final TXStateInterface tx) {
+
+    /** clear the TXContext and remove from global list when thread closes */
+    public final void threadClose() {
+      clearTXStateAll();
```
Wondering if it should just be cleared or rolledback?

sumwale replied:
Had thought about it but was a bit circumspect if it could mean deadlocks or other issues given that other pools would also have been shutdown.

sumwale replied:
@suranjan Ok, pushed a change to rollback here if cache is not closing -- please check

Commit:
make naming of threads consistent

Commit:
review comment: rollback TXState in thread close if cache is not closing

suranjan commented:
Looks good to me.

Commit:
masquerade transaction in rollback for cache close

removed clear of static hosted TXContext map since the same threads can be re-used
on reconnect (e.g. Spark executor threads, RMI threads)

added logging at TX.VERBOSE for clearTXStateAll
